src/xrpld/app/tx/detail/LoanBrokerCoverClawback.cpp

```diff
@@ -0,0 +1,336 @@
+        NumberRoundModeGuard mg1(Number::upward);
+        auto const minRequiredCover = tenthBipsOfValue(
+            sleBroker[sfDebtTotal], TenthBips32(sleBroker[sfCoverRateMinimum]));
+        // The subtraction probably won't round, but round down if it does.
+        NumberRoundModeGuard mg2(Number::downward);
+        return sleBroker[sfCoverAvailable] - minRequiredCover;
+    }();
+    if (maxClawAmount <= beast::zero)
+        return Unexpected(tecINSUFFICIENT_FUNDS);
+
+    // Use the vaultAsset here, because it will be the right type in all
+    // circumstances. The amount may be an IOU indicating the pseudo-account's
+    // asset, which is correct, but not what is needed here.
+    if (!amount || *amount == beast::zero)
+        return STAmount{vaultAsset, maxClawAmount};
+    Number const magnitude{*amount};
+    if (magnitude > maxClawAmount)
+        return STAmount{vaultAsset, maxClawAmount};
+    return STAmount{vaultAsset, magnitude};
+}
+
+template <ValidIssueType T>
+static TER
+preclaimHelper(
+    PreclaimContext const& ctx,
+    SLE const& sleIssuer,
+    STAmount const& clawAmount);
+
+template <>
+TER
+preclaimHelper<Issue>(
+    PreclaimContext const& ctx,
+    SLE const& sleIssuer,
+    STAmount const& clawAmount)
+{
+    // If AllowTrustLineClawback is not set or NoFreeze is set, return no
+    // permission
+    if (!(sleIssuer.isFlag(lsfAllowTrustLineClawback)) ||
+        (sleIssuer.isFlag(lsfNoFreeze)))
+        return tecNO_PERMISSION;
+
+    return tesSUCCESS;
+}
+
+template <>
+TER
+preclaimHelper<MPTIssue>(
+    PreclaimContext const& ctx,
+    SLE const& sleIssuer,
+    STAmount const& clawAmount)
+{
+    auto const issuanceKey =
+        keylet::mptIssuance(clawAmount.get<MPTIssue>().getMptID());
+    auto const sleIssuance = ctx.view.read(issuanceKey);
+    if (!sleIssuance)
+        return tecOBJECT_NOT_FOUND;
+
+    if (!sleIssuance->isFlag(lsfMPTCanClawback) ||
+        !sleIssuance->isFlag(lsfMPTCanLock))
+        return tecNO_PERMISSION;
+
+    // With all the checking already done, this should be impossible
+    if (sleIssuance->at(sfIssuer) != sleIssuer[sfAccount])
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    return tesSUCCESS;
+}
+
+TER
+LoanBrokerCoverClawback::preclaim(PreclaimContext const& ctx)
+{
+    auto const& tx = ctx.tx;
+
+    auto const account = tx[sfAccount];
+    auto const findBrokerID = determineBrokerID(ctx.view, tx);
+    if (!findBrokerID)
+        return findBrokerID.error();
+    auto const brokerID = *findBrokerID;
+    auto const amount = tx[~sfAmount];
+
+    auto const sleBroker = ctx.view.read(keylet::loanbroker(brokerID));
+    if (!sleBroker)
+    {
+        JLOG(ctx.j.warn()) << "LoanBroker does not exist.";
+        return tecNO_ENTRY;
+    }
+
+    auto const brokerPseudoAccountID = sleBroker->at(sfAccount);
+
+    auto const vault = ctx.view.read(keylet::vault(sleBroker->at(sfVaultID)));
+    if (!vault)
+        return tecINTERNAL;
```

Comment:
Elsewhere we return `tefBAD_LEDGER`

```suggestion
    {
        // LCOV_EXCL_START
        JLOG(ctx.j.fatal()) << "Vault is missing for Broker " << brokerID;
        return tefBAD_LEDGER;
        // LCOV_EXCL_STOP  
    }

```